app/ui/hooks/useFetchComments.ts

```diff
@@ -0,0 +1,73 @@
+import {Dispatch, SetStateAction, useCallback, useEffect, useState} from "react";
+import {IComment, ITweet} from "@/app/lib/definitions";
+import useQuery, {IQueryParams} from "@/app/ui/hooks/useQuery";
+import useNotification from "@/app/ui/hooks/useNotification";
+import {NotificationType} from "@/app/ui/context/NotificationContext";
+import {switchMap} from "rxjs";
+
+export default function useFetchComments(tweet: ITweet, setTweet: Dispatch<SetStateAction<ITweet | undefined>>) {
+
+    const [comments, setComments] = useState<IComment[]>([])
+    const query$ = useQuery()
+    const {setNotification} = useNotification()
+
+    const fetchComments = useCallback(() => {
+        const params: IQueryParams = {
+            URL: `/tweets/${tweet.id}/comments`,
+            method: 'GET',
+            authorized: true,
+        }
+
+        query$<{ comments: IComment[] }>(params).subscribe(({data}) => {
+            setComments(data.comments)
+        })
+    }, [query$, tweet])
+
+    const notifyError = useCallback(() => {
+        setNotification({
+            type: NotificationType.ERROR,
+            message: "An error occurred."
+        })
+    }, [setNotification])
+
+    const comment = useCallback((comment: FormData) => {
+        const postParams: IQueryParams = {
+            URL: `/tweets/${tweet.id}/comments`,
+            method: 'POST',
+            authorized: true,
+            body: comment
+        }
+
+        const getParams: IQueryParams = {
+            URL: `/tweets/${tweet.id}/comments`,
+            method: 'GET',
+            authorized: true,
+        }
+        query$<{}>(postParams)
+            .pipe(
+                switchMap(() => {
+                    setTweet(tweet => tweet && {
+                        ...tweet,
+                        _count: {...tweet._count, comments: tweet._count.comments + 1}
+                    })
+                    return query$<{ comments: IComment[] }>(getParams)
```

Comment:
Quick note: when we are adding a new entry to the database, for example a comment, there can be a lot more in the database. Right now this won't be a problem, but when you are working on a bigger system it is a waste of resource to re-fetch the comments. This is when we are being optimistic:
- When you use a CREATE/POST operation, we check the status code of the response (here, the subscribe's next parameter, I'm sure you've seen it in the code before)
- If the status code is created/ok, we believe that the entry has been added
- Therefore, for one comment, we do not fetch the database again
- Instead, just add the new comment to the array state, we are OPTIMISTIC about it, that it was added. (This will be react's new feature too, the useOptimistic hook)